File: ant-sdk/antd-zig/examples/04-files.zig
const std = @import("std");
const antd = @import("antd");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var client = antd.Client.init(allocator, antd.default_base_url);
    defer client.deinit();

    const tmp = "/tmp/antd-zig-04-files";
    std.fs.deleteTreeAbsolute(tmp) catch {};
    try std.fs.makeDirAbsolute(tmp);
    defer std.fs.deleteTreeAbsolute(tmp) catch {};

    const file_content = "Hello from a file on Autonomi!";

    const src_file = tmp ++ "/hello.txt";
    {
        const f = try std.fs.createFileAbsolute(src_file, .{});
        defer f.close();
        try f.writeAll(file_content);
    }

    const est = try client.fileCost(src_file, true, .auto);
    defer est.deinit(allocator);
    std.debug.print(
        "Estimate: {d} bytes in {d} chunks, storage {s} atto, gas {s} wei, mode {s}\n",
        .{ est.file_size, est.chunk_count, est.cost, est.estimated_gas_cost_wei, est.payment_mode },
    );

    const upload_result = try client.filePutPublic(src_file, .auto);
    defer upload_result.deinit(allocator);

    std.debug.print("File uploaded at: {s}\n", .{upload_result.address});
    std.debug.print("Storage cost: {s} atto, gas: {s} wei\n", .{ upload_result.storage_cost_atto, upload_result.gas_cost_wei });
    std.debug.print("Chunks stored: {d}, payment mode: {s}\n", .{ upload_result.chunks_stored, upload_result.payment_mode_used });

    const dst_file = tmp ++ "/hello.txt.downloaded";
    try client.fileGetPublic(upload_result.address, dst_file);
    std.debug.print("File downloaded to {s}\n", .{dst_file});

    {
        const got = try std.fs.cwd().readFileAlloc(allocator, dst_file, 1 << 20);
        defer allocator.free(got);
        if (!std.mem.eql(u8, got, file_content)) {
            std.debug.print("round-trip mismatch on hello.txt\n", .{});
            return error.RoundTripMismatch;
        }
    }

    std.debug.print("File upload/download OK!\n", .{});
}

File: ant-sdk/antd-lua/examples/04-files.lua
--- Example: Upload and download files publicly, with round-trip assertions.

local antd = require("antd")

local client = antd.new_client()

local function write_file(path, content)
    local f = assert(io.open(path, "wb"))
    f:write(content)
    f:close()
end

local function read_file(path)
    local f = assert(io.open(path, "rb"))
    local content = f:read("*a")
    f:close()
    return content
end

local function run_cmd(cmd)
    local ok = os.execute(cmd)
    return ok == true or ok == 0
end

local tmp = "/tmp/antd-lua-04-files"
run_cmd("rm -rf " .. tmp)
assert(run_cmd("mkdir -p " .. tmp))

local file_content = "Hello from a file on Autonomi!"

local src_file = tmp .. "/hello.txt"
write_file(src_file, file_content)

local cost, err5 = client:file_cost(src_file, true)
if err5 then
    print("Cost error: " .. err5.message)
    os.exit(1)
end
print(string.format("Estimated upload cost: %s atto (%d chunks)", cost.cost, cost.chunk_count))

local result, err = client:file_put_public(src_file)
if err then
    print("Upload error: " .. err.message)
    os.exit(1)
end

print("File uploaded at: " .. result.address)
print("Storage cost: " .. result.storage_cost_atto .. " atto, gas: " .. result.gas_cost_wei .. " wei")
print("Chunks stored: " .. result.chunks_stored .. ", payment mode: " .. result.payment_mode_used)

local dst_file = tmp .. "/hello.txt.downloaded"
local _, err2 = client:file_get_public(result.address, dst_file)
if err2 then
    print("Download error: " .. err2.message)
    os.exit(1)
end

print("File downloaded to " .. dst_file)

if read_file(dst_file) ~= file_content then
    run_cmd("rm -rf " .. tmp)
    print("Round-trip mismatch on hello.txt")
    os.exit(1)
end

run_cmd("rm -rf " .. tmp)
print("File upload/download OK!")
